### PR TITLE
refactor and abstract ssn validation with form builder

### PR DESCRIPTION
**Pull request by scaustin34**
# PR Checklist

Please check if your PR fulfills the following requirements
- [x] The title follows our [guidelines](https://github.com/ideacrew/enroll/blob/trunk/CONTRIBUTING.md#commit
)
- [ ] Tests for the changes have been added (for bugfixes / features)

# PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature (requires Feature flag)
- [ ] Data fix or migration (inert code, no impact until run)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Dependency updates (e.g., add a new gem or update version)

# What is the ticket # detailing the issue?

Ticket: 
https://www.pivotaltracker.com/n/projects/2648255/stories/185808202
https://www.pivotaltracker.com/n/projects/2648255/stories/185941385

# A brief description of the changes

Current behavior: SSN fields do not have front end validation

New behavior: SSN fields do have front end validation that prevents invalid social security numbers from being submitted

# Feature Flag

For all new feature development, a feature flag is required to control the exposure of the feature to our end users. A feature flag needs a corresponding environment variable that is used to initialize the state of the flag. Please share the name of the environment variable below that would enable/disable the feature and which client(s) it applies to.

Variable name: `SSN_UI_VALIDATION_IS_ENABLED`

- [ ] DC
- [x] ME

### CODE REVIEW

File: app/builders/hbx_form_builder.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+      placeholder: "#{l10n('social_security').to_s.upcase} #{'*' if is_required}",
+      readonly: is_readonly
+    }
+    if EnrollRegistry.feature_enabled?(:ssn_ui_validation)
+      @template.text_field(
+        @object_name,
+        method,
+        options.merge(
+          pattern: "(?!666|000|9\\d{2})\\d{3}[\\- ]{0,1}(?!00)\\d{2}[\\- ]{0,1}(?!0{4})\\d{4}",
+          oninvalid: "this.setCustomValidity('Invalid Social Security number.')",
```
Comment (scaustin34) [inline app/builders/hbx_form_builder.rb:24]: probably should throw this string into the translation files, while we're here.

Comment (adamrunner) [reply app/builders/hbx_form_builder.rb:24]: That's a good idea.

**Review by TreyE — CHANGES_REQUESTED** [review]
A few issues here:
1. The PR checklist isn't complete.  Please fill it out with a description of what you are changing - especially the ticket I'm meant to refer back to.
2. I'm not seeing any specs that test specific allowed patterns.  Since we've introduced a new class I'd expect to see at the very least a spec file for that new form builder that somehow verifies some known patterns for acceptable vs disallowed SSNs (checking for some of the fun little combinations of things like if the user SSN input comes in with a dash, if it doesn't, and whatever other fun stuff could happen based on how the javascript frontend is feeling today).
3. Disabling the Rubocop documentation cop globally makes me nervous.  I'd recommend finding out exactly what it was complaining about and fixing that - it's here to help us get documentation and better practices around what is an old code base with a lot of idiosyncrasies in the event we refactor.
4. There's no dummy file for the new form builder in the financial assistance engine - this one gets its own section. 

## A dummy what now?

This is a particular code smell I see here that I'm going to do my best to explain - and it's not necessarily an issue with *your code* - it's more that your code has maybe uncovered an area that isn't properly protected by specs or has other issues.

With the way we use engines, when the specs for them are run in isolation, they aren't supposed to be able to access constants or classes hosted in the main app - for example your class living in `app/builders/hbx_form_builder.rb`.  The way we get around this is we put a 'dummy' copy of the class into the 'spec/dummy/app/whatever' directory of the engine.  In this case it would be in `components/financial_assistance/spec/dummy/app/builders/hbx_form_builder`.  Typically if you don't do this, specs run for the engine will start throwing fun little constant missing errors.

The fact that we replaced the base form builder, there isn't a dummy file in the financial assistance engine, and specs aren't failing badly makes me worried.  It makes me think there are possibly two different weird things going on here:
1. The financial assistance engine isn't actually configured to/isn't actually using your form builder - otherwise it should be complaining that it doesn't know what class you are talking about when the engine-level specs run.
2. Even if the form builder is getting configured correctly in the engine level and for the engine-level specs - I suspect we are lacking specs covering the actual forms themselves.  The fact that both no specs are failing at the engine level *and* you didn't have to change/fix any specs in the engine for the two views we altered makes me think this is the case.

So, how do we fix this?
1.  Make sure specs for the _edit.html and _dependent_form.html templates exist.  If they don't, we will need them.
2.  Regardless of if the specs exist or not, we need to verify that the new form builder you created is being used as the base class when the engine is run in isolation.  We need to understand what is happening - it's weird if it's using your new form builder and there are specs, and they aren't crashing.  If there aren't specs, writing them will help verify your base class is being used.  If this is working right, your specs should keep failing until you copy that file into the spec/dummy directory.

File: .rspec [hunk 2]
```diff
@@ -2,4 +2,4 @@
 --require rails_helper
 --require rspec/instafail
 --format RSpec::Instafail
---format progress
+--format documentation
```
[the remark this answers is not in the document]
Comment (adamrunner) [reply .rspec:5]: I talked to no one about this change, it was intended to be a local only change, I'll change this back to what it was.

File: Gemfile [hunk 3]
```diff
@@ -113,47 +106,46 @@ gem 'wicked_pdf',               '~> 1.1.0'
...
   gem 'web-console',            '>= 3'
   gem 'listen',                 '>= 3.0.5', '< 3.2'
-
-  # Spring speeds up development by keeping your application running in the background. Read more: https://github.com/rails/spring
   gem 'spring'
   gem 'spring-watcher-listen'
 end
 
 group :development, :test do
+  gem 'pry-rails'
```
[the remark this answers is not in the document]
Comment (adamrunner) [reply Gemfile:127]: That's great to know - I'm sure I've ran into sluggishness in loading test environments before and I've never attributed it to `pry`.

File: .rubocop.yml [hunk 4]
```diff
@@ -108,9 +108,8 @@ Style/AndOr:
 Style/BlockComments:
   Enabled: false
 
-# We will want to turn this back on or customize it more fully
 Style/Documentation:
-  Enabled: true
+  Enabled: false
```
[the remark this answers is not in the document]
Comment (adamrunner) [reply .rubocop.yml:112]: @TreyE Ah again, a local change that got committed erroneously. My bad. 

Though, I'd love to have a discussion about the behavior that this particular rubocop encourages - I'm not sure if it's adding the value we want it to.

File: spec/views/shared/person/_personal_information.html.erb_spec.rb [hunk 5]
```diff
@@ -23,6 +23,6 @@
     expect(rendered).to have_selector('input[placeholder="FIRST NAME *"]')
     expect(rendered).to have_selector('input[placeholder="LAST NAME *"]')
     expect(rendered).to have_selector('input[placeholder="DATE OF BIRTH *"]')
-    expect(rendered).to have_selector('input[placeholder="SOCIAL SECURITY"]')
+    expect(rendered).to have_selector('input[placeholder="SOCIAL SECURITY *"]')
```
[the remark this answers is not in the document]
Comment (adamrunner) [reply spec/views/shared/person/_personal_information.html.erb_spec.rb:26]: More specifically, in this case the SSN field was previously required, but the placeholder text did not have the `*` appended to it - when we switched to the replacement form builder, the test was failing but the test was incorrect in the first place. This change makes the assertion behave as expected.

SSN is not always required, you can pass a `required` attribute to the `ssn_field` method, which should conditionally append the `*` character to the placeholder value.

File: app/views/insured/family_members/_dependent_form.html.erb [hunk 6]
```diff
@@ -105,13 +105,7 @@
           <% unless @person.has_active_resident_role? %>
             <div class="col-md-3 col-sm-3 no-pd">
               <div class="form-group form-group-lg">
-                <% if EnrollRegistry.feature_enabled?(:ssn_ui_validation) %>
-                  <%= f.text_field :ssn, value: dependent.ssn,class: "home required floatlabel form-control mask-ssn",
-                    pattern: "(?!666|000|9\\d{2})\\d{3}[\\- ]{0,1}(?!00)\\d{2}[\\- ]{0,1}(?!0{4})\\d{4}", oninvalid: "this.setCustomValidity('Invalid Social Security number.')",
-                    oninput: "this.setCustomValidity('')", placeholder: "#{l10n("social_security").to_s.upcase}" %>
-                <% else %>
-                  <%= f.text_field :ssn, value: dependent.ssn,class: "home required floatlabel form-control mask-ssn", placeholder: "#{l10n("social_security").to_s.upcase}" %>
-                <% end %>
```
[the remark this answers is not in the document]
Comment (adamrunner) [reply app/views/insured/family_members/_dependent_form.html.erb:114]: @TreyE 

Are there more needed than the following:

```
db/seedfiles/translations/en/cca/insured.rb
db/seedfiles/translations/en/dc/insured.rb
db/seedfiles/translations/en/me/insured.rb
```

**Pull request closed without merging**